### PR TITLE
Warn the user when unverified user has changed their identity

### PR DESCRIPTION
**Pull request by bmarty**
## Content

As per https://github.com/element-hq/element-meta/issues/2544

Using https://github.com/matrix-org/matrix-rust-sdk/pull/4022 and https://github.com/matrix-org/matrix-rust-sdk/pull/4068

Draft since the application is crashing for now when clicking on OK to pin the user. **EDIT** The crash has been fixed in https://github.com/matrix-org/matrix-rust-sdk/pull/4089 so an SDK release will be neeeded, but this is not blocking this PR anymore (pinning identity cannot be tested, but the PR can be reviewed).

## Motivation and context

https://github.com/element-hq/element-meta/issues/2544

## Tests

- Alice and Bob in a e2e room
- Alice has not verified Bob
- Messages can be sent
- Bob reset their identity
- Alice should see the banner, until they click on OK to pin the identity of Bob (but currently this is crashing)

## Tested devices

- [ ] Physical
- [x] Emulator
- OS version(s):

## Checklist

- [ ] Changes have been tested on an Android device or Android emulator with API 23
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/element-hq/element-x-android/blob/develop/CONTRIBUTING.md#accessibility
- [ ] Pull request is based on the develop branch
- [ ] Pull request title will be used in the release note, it clearly define what will change for the user
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [ ] You've made a self review of your PR

### CODE REVIEW

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/MessagesNode.kt [hunk 1]
```diff
@@ -135,16 +135,20 @@ class MessagesNode @AssistedInject constructor(
                 callbacks.forEach { it.onUserDataClick(permalink.userId) }
             }
             is PermalinkData.RoomLink -> {
-                handleRoomLinkClick(permalink, eventSink)
+                handleRoomLinkClick(activity, permalink, eventSink)
             }
             is PermalinkData.FallbackLink,
             is PermalinkData.RoomEmailInviteLink -> {
-                context.openUrlInExternalApp(url)
+                activity.openUrlInChromeCustomTab(null, darkTheme, url)
```
Comment (bmarty) [inline features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/MessagesNode.kt:142]: Note: at some point, direct usage of `openUrlInExternalApp` should be forbidden, but this is out of scope of this PR.

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/crypto/identity/IdentityChangeStatePresenter.kt [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import javax.inject.Inject
+
+class IdentityChangeStatePresenter @Inject constructor(
+    private val room: MatrixRoom,
+) : Presenter<IdentityChangeState> {
+    @Composable
+    override fun present(): IdentityChangeState {
+        val roomMemberIdentityStateChange = remember {
```
Comment (jmartinesp) [inline features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/crypto/identity/IdentityChangeStatePresenter.kt:42]: Couldn't this just use `produceState`?

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/crypto/identity/IdentityChangeStateView.kt [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+import io.element.android.libraries.matrix.ui.model.getAvatarData
+import io.element.android.libraries.ui.strings.CommonStrings
+
+@Composable
+fun IdentityChangeStateView(
+    state: IdentityChangeState,
+    modifier: Modifier = Modifier,
+) {
+    // Pick the first identity change to PinViolation
+    val identityChange = state.roomMemberIdentityStateChanges.firstOrNull {
```
Comment (jmartinesp) [inline features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/crypto/identity/IdentityChangeStateView.kt:35]: Maybe rename it to something like `dangerousIdentityChange` or `pinViolationIdentityChange`?

**Review by jmartinesp — APPROVED** [review]
LGTM, just a couple of suggestions.

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/crypto/identity/IdentityChangeStatePresenter.kt (hunk 2 above)
Comment (ganfra) [reply features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/crypto/identity/IdentityChangeStatePresenter.kt:42]: Yes this would make more sense. Also use `by` delegate when possible?

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/crypto/identity/IdentityChangeStatePresenter.kt [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                is IdentityChangeEvent.Submit -> {
+                    ignoredUserIdChange.value += event.userId
+                    // TODO notify the SDK
+                }
+            }
+        }
+
+        return IdentityChangeState(
+            roomMemberIdentityStateChanges = roomMemberIdentityStateChange.value
+                .filter { it.roomMember.userId !in ignoredUserIdChange.value }
```
Comment (ganfra) [inline features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/crypto/identity/IdentityChangeStatePresenter.kt]: I guess the list will never be large, but we are filtering on the main thread for every composition ^

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/crypto/identity/IdentityChangeState.kt [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import io.element.android.libraries.matrix.api.room.RoomMember
+import kotlinx.collections.immutable.ImmutableList
+
+data class IdentityChangeState(
+    val roomMemberIdentityStateChanges: ImmutableList<RoomMemberIdentityStateChange>,
+    val eventSink: (IdentityChangeEvent) -> Unit,
+)
+
+data class RoomMemberIdentityStateChange(
+    val roomMember: RoomMember,
```
Comment (ganfra) [inline features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/crypto/identity/IdentityChangeState.kt:20]: Could we create a specific data class instead of passing `RoomMember` if we only need a few fields?

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/crypto/identity/IdentityChangeStatePresenter.kt (hunk 4 above)
Comment (bmarty) [reply features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/crypto/identity/IdentityChangeStatePresenter.kt]: This code does not exist anymore, are you reviewing an old version?

**Comment by bmarty** [comment]
@ganfra @jmartinesp thanks for the fast review. I think I have handled all your remarks.
I took the opportunity to update the code regarding typing notification as it had similar problems.
